Set BuildKit version using buildinfo

crazy-max opened the pull request:
follow-up https://github.com/moby/moby/pull/44686#discussion_r1094416821
closes #44917

**- What I did**

This sets BuildKit version from the build information embedded in running binary so we are aligned with the expected vendoring.

**- How I did it**

We iterate over all dependencies and find the BuildKit one and set the right version. We also check if the module is replaced and use it in this case.

There is also additional checks if a pseudo version is detected (see comments in code for more info)

But there is a drawback...

As we are not supporting (yet) Go modules, dependencies are not embedded in the binary and therefore we can't retrieve the BuildKit module version. Last commit enables Go modules when building dockerd and docker-proxy so we have buildinfo with deps embedded in the binary. Once we switch to Go modules, we can remove this workaround.

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

crazy-max committed:
hack: enable Go modules when building dockerd and docker-proxy

This is a workaround to have buildinfo with deps embedded in the
binary. We need to create a go.mod file before building with
-modfile=vendor.mod, otherwise it fails with:
"-modfile cannot be used to set the module root directory."

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

tonistiigi reviewed (commented):
I like this. If it is not too hard, could we have a test for this?

crazy-max commented:
> I like this. If it is not too hard, could we have a test for this?

Added some tests.

thaJeztah commented on builder/builder-next/worker/worker.go:
```diff
@@ -523,3 +534,57 @@ type emptyProvider struct {
+			// if the version is replaced, get the replaced version
+			modVersion = dep.Replace.Version
+		}
+		if module.IsPseudoVersion(modVersion) {
+			// if the version is a pseudo version, get the base version
+			// e.g. v0.10.7-0.20230306143919-70f2ad56d3e5 => v0.10.6
+			if base, err := module.PseudoVersionBase(modVersion); err == nil && base != "" {
+				// set canonical version of the base version (removes +incompatible suffix)
+				// e.g. v2.1.2+incompatible => v2.1.2
+				base = semver.Canonical(base)
+				// if the version is a pseudo version, get the revision
+				// e.g. v0.10.7-0.20230306143919-70f2ad56d3e5 => 70f2ad5
+				if rev, err := module.PseudoVersionRev(modVersion); err == nil && rev != "" {
+					// append the revision to the version
+					// we set the short commit sha instead of the uncommon
+					// one (https://github.com/golang/go/issues/34745)
+					// e.g. v0.10.7-0.20230306143919-70f2ad56d3e5 => v0.10.6+70f2ad5
+					modVersion = base + "+" + rev[:7]
+				} else {
+					// if the revision is not available, use the base version
+					modVersion = base
+				}
+			}
+		}
```
Oh, double check my suggestion, looks like I may have slipped up something 🤔 

```
=== RUN   TestModuleVersion/returns_the_base_version_of_pseudo_version
    worker_test.go:64: moduleVersion() = v0.10.6, want v0.10.6+70f2ad5
```

crazy-max commented:
> I'm curious; some users asked about this; "what dockerfile syntax version is the default?", and I'm wondering if we should surface that somewhere. Is that information somehow captured in BuildKit, or would that require `frontend/dockerfile` to become its own module?

I have done this in a PR on BuildKit repo: https://github.com/moby/buildkit/pull/2725#issuecomment-1068205802

See also the context of why we have not merged this change: https://github.com/moby/buildkit/pull/2725#discussion_r828235303

Here is the version constraint: https://github.com/moby/buildkit/blob/43ad8abf96907b88fc7dbcd44c6fc532a19227f0/frontend/dockerfile/release/releases.json

cpuguy83 commented on builder/builder-next/worker/worker.go:
```diff
@@ -46,11 +48,20 @@ import (
+
 func init() {
-	version.Version = "v0.11.0-rc3"
+	if bkversion := moduleVersion(readBuildInfo()); bkversion != "" {
```
Can we set this when requested rather than at init?

cpuguy83 reviewed (commented):
As much as I dislike having to use ldflags to set vars... I'm not sure how I feel about this either.

thaJeztah commented:
> As much as I dislike having to use ldflags to set vars... I'm not sure how I feel about this either.

Yeah, I somewhat agree (at least currently). I don't mind exploring options, but perhaps we should not rush it.

That said; given that this version depends on the vendor, can we instead (at least for now); 

- update `vendor.sh` to `sed/` https://github.com/moby/moby/blob/61656464d86a033c4b2b19b053d93c2f0035389f/builder/builder-next/worker/worker.go#L52-L54
- ^^ I honestly forgot we had that hard-coded, and CI didn't validate it ⚠️ 
- validate that in CI
- and perhaps in `hack/make.sh build` (if anyone would build with updated versions in `vendor.mod` without using `hack/vendor.sh`)

Perhaps "somewhat ugly", but given that we also commit `vendor.mod` and the actual vendoring in git, I don't think it's far from the truth in that respect.

cpuguy83 commented:
We could use "go generate" to generate the version.
The generated file could either be a proper or even just plain text that we use "go:embed" with.

tonistiigi commented on builder/builder-next/worker/worker.go:
```diff
@@ -46,11 +47,50 @@ import (
 func init() {
-	version.Version = "v0.11.0-rc3"
+	// returns the build information embedded in running binary
+	if bi, ok := debug.ReadBuildInfo(); ok {
```
https://github.com/moby/moby/pull/45304/files/f8d3c2d7208915c96c9424e250aadbb3871ea680#diff-cea55065cd7f6cf43128f33bb9ad7b223c244fa6ad23a9a3852f9981b3d71e64

crazy-max committed:
Set BuildKit version using buildinfo

This sets BuildKit version from the build information embedded
in running binary so we are aligned with the expected vendoring.

We iterate over all dependencies and find the BuildKit one
and set the right version. We also check if the module is
replaced and use it this case.

There is also additional checks if a pseudo version is
detected. See comments in code for more info.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

tianon commented:
I definitely like this approach a lot better than https://github.com/moby/moby/pull/44917 :sweat_smile:

Honestly the only thing here that slightly bothers me is caching the full buildinfo response instead of just the buildkit version, but it's probably small enough to be mostly fine?

Can you elaborate on your concerns with reading the buildinfo, @cpuguy83? :bow: :pray:

corhere commented on pkg/mod/mod.go:
```diff
@@ -0,0 +1,67 @@
+)
+
+var (
+	buildInfoOnce = sync.Once{}
```
```suggestion
	buildInfoOnce   sync.Once
```

corhere commented on pkg/mod/mod.go:
```diff
@@ -0,0 +1,67 @@
+	buildInfo     *debug.BuildInfo
+)
+
+func GetVersion(name string) (modVersion string) {
```
```suggestion
func Version(name string) (modVersion string) {
```

cpuguy83 commented on builder/builder-next/worker/worker.go:
```diff
@@ -50,7 +51,9 @@ import (
 
 func init() {
-	version.Version = "v0.11.6"
+	if v := mod.GetVersion("github.com/moby/buildkit"); v != "" {
```
I'd still really like to avoid doing this in a `func init()`

corhere commented on pkg/mod/mod.go:
```diff
@@ -0,0 +1,67 @@
+			// e.g. v0.10.7-0.20230306143919-70f2ad56d3e5 => 70f2ad5
+			if rev, err := module.PseudoVersionRev(modVersion); err == nil && rev != "" {
+				// append the revision to the version
+				// we set the short commit sha instead of the uncommon
+				// one (https://github.com/golang/go/issues/34745)
+				// e.g. v0.10.7-0.20230306143919-70f2ad56d3e5 => v0.10.6+70f2ad5
+				modVersion = base + "+" + rev[:7]
```
Without actually consulting the Git repository, there's not enough information to know whether seven characters is long enough to unambiguously identify a single commit. I have personally been burned by a situation where a repo grew large enough that seven characters was insufficient. Please do not truncate the commit hash further than Go modules already do.

corhere commented on pkg/mod/mod.go:
I don't think this package should be made available for import outside the repo. The handling of pseudo-versions is rather subtle and opinionated, and might not be appropriate to use elsewhere. And we'd have to maintain backwards compatibility for as long as the package is named `"github.com/docker/docker/pkg/mod"` (read: until we migrate this repo to Go modules). Please move this package under the `internal` tree. We can always promote this to an externally importable package later, but we can't do the reverse.

neersighted commented on hack/make/.binary:
```diff
@@ -62,11 +62,19 @@ source "${MAKEDIR}/.go-autogen"
 		set -x
 	fi
-	go build -o "$DEST/$BINARY_FULLNAME" "${BUILDFLAGS[@]}" -ldflags "$LDFLAGS $LDFLAGS_STATIC $DOCKER_LDFLAGS" ${GO_PACKAGE}
+	GO111MODULE=on go build -mod=vendor -modfile=vendor.mod -o "$DEST/$BINARY_FULLNAME" "${BUILDFLAGS[@]}" -ldflags "$LDFLAGS $LDFLAGS_STATIC $DOCKER_LDFLAGS" ${GO_PACKAGE}
```
I'm late here, but was there a reason not to prefer the `with-go-mod` wrapper script we already have?

crazy-max replied:
Log output was a bit clunky using `with-go-mod` and also the dependency on `git` was an issue as we ignore `.git` in `.dockerignore`: https://github.com/moby/moby/pull/44917/files#diff-fa876a8e859e8583907d556eeed2e4b8cc98d7847b2124521aaef1e9984f50cf
